### PR TITLE
Add emformer hubert model architecture

### PR DESCRIPTION
**Pull request by nateanl**

### CODE REVIEW

File: torchaudio/prototype/models/_emformer_hubert.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
+from typing import List, Optional, Tuple
+
+import torch
+from torchaudio.models import Wav2Vec2Model
+from torchaudio.models.emformer import Emformer
+from torchaudio.models.rnnt import _TimeReduction
+
+
+class FeatureExtractor(torch.nn.Module):
```
Comment (carolineechen) [inline torchaudio/prototype/models/_emformer_hubert.py:9]: Thoughts on renaming to FeatureEncoder, similar to Conformer W2V2?

File: torchaudio/prototype/models/_emformer_hubert.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+def emformer_hubert_base(
+    extractor_input_dim: int = 80,
+    extractor_output_dim: int = 128,
+    encoder_dropout: float = 0.1,
+) -> Wav2Vec2Model:
+    """Build Emformer HuBERT Model with 20 Emformer layers.
+
+    Args:
+        extractor_input_dim (int): The input dimension for feature extractor.
```
Comment (carolineechen) [inline torchaudio/prototype/models/_emformer_hubert.py:296]: similarly for the other params
```suggestion
        extractor_input_dim (int, optional): The input dimension for feature extractor. (Default: 80)
```

File: torchaudio/prototype/models/_emformer_hubert.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+            input (torch.Tensor): The input feature for emformer encoder.
+                Tensor with dimensions `(batch, time, feature_dim)`.
+            lengths (torch.Tensor or None): Valid length of each input sample.
+                Tensor with dimension `(batch, )`.
+            num_layers (int or None, optional): If not ``None``, returns the first
+                `num_layers` layers of Tensors as the output, otherwise returns the
+                Tensors from all emformer layers.
+
+        Returns:
+            (List[torch.Tensor]):
```
Comment (hwangjeff) [inline torchaudio/prototype/models/_emformer_hubert.py:120]:
```suggestion
            List[torch.Tensor]:
```

File: torchaudio/prototype/models/_emformer_hubert.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    ) -> torch.Tensor:
+        """
+        Args:
+            input (torch.Tensor): The input feature for emformer encoder.
+                Tensor with dimensions `(batch, time, feature_dim)`.
+            lengths (torch.Tensor or None): Valid length of each input sample.
+                Tensor with dimension `(batch, )`.
+
+        Returns:
+            (torch.Tensor): The feature Tensor after emformer encoder.
```
Comment (hwangjeff) [inline torchaudio/prototype/models/_emformer_hubert.py:89]:
```suggestion
            torch.Tensor: The feature Tensor after emformer encoder.
```

File: test/torchaudio_unittest/prototype/emformer_hubert_test.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
+import torch
+from parameterized import parameterized
+from torchaudio.prototype.models import emformer_hubert_base
+from torchaudio_unittest.common_utils import skipIfNoCuda, torch_script, TorchaudioTestCase
+
+
+class TestEmformerHuBERT(TorchaudioTestCase):
```
Comment (carolineechen) [inline test/torchaudio_unittest/prototype/emformer_hubert_test.py:7]: do you think we could just reuse this tests file and use parameterized to include both conformerw2v2 and emformerhubert? https://github.com/pytorch/audio/blob/main/test/torchaudio_unittest/prototype/conformer_wav2vec2_test.py

Comment (nateanl) [reply test/torchaudio_unittest/prototype/emformer_hubert_test.py:7]: i think so, will update the existing test then

Comment (nateanl) [reply test/torchaudio_unittest/prototype/emformer_hubert_test.py]: I renamed the test file to `ssl_model_test.py` to reflect the change, let me know if you have a better idea.

File: torchaudio/prototype/models/_emformer_hubert.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
+from typing import List, Optional, Tuple
+
+import torch
+from torchaudio.models import Wav2Vec2Model
+from torchaudio.models.emformer import Emformer
+from torchaudio.models.rnnt import _TimeReduction
+
+
+class FeatureEncoder(torch.nn.Module):
+    """Extract features from log mel-spectrogram input. Consists of linear layer and time reduction layer.
```
Comment (carolineechen) [inline torchaudio/prototype/models/_emformer_hubert.py:10]: nit: I normally see `logmel spectrogram` or `log-mel spectrogram` and not `log mel-spectrogram`?

Comment (nateanl) [reply torchaudio/prototype/models/_emformer_hubert.py:10]: that's right, will change it to log-mel spectrogram.

**Review by carolineechen — APPROVED** [review]

File: test/torchaudio_unittest/prototype/ssl_model_test.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            high=num_frames,
+            size=[
+                batch_size,
+            ],
+            device=device,
+        )
+
+        model(features, lengths)
+
+    @parameterized.expand(
```
Comment (carolineechen) [inline test/torchaudio_unittest/prototype/ssl_model_test.py:25]: nit: can use @nested_params here

Comment (nateanl) [reply test/torchaudio_unittest/prototype/ssl_model_test.py:25]: sure, will address in a separate pr

**Pull request closed without merging**